File: gitlab_watch.py
#!/usr/bin/env python3
"""
gitlab_watch.py — live terminal feed of every team member's GitLab activity.

Talks to the GitLab REST API directly (no Claude MCP needed), so it runs
anywhere the GitLab server is reachable (office WiFi / VPN).

Config comes from environment variables (or a .env file next to this script):

  GITLAB_URL       Base URL of your GitLab, e.g. https://gitlab.example.com   (required)
  GITLAB_TOKEN     Personal access token with `read_api` scope                (required)
  GITLAB_GROUP     Group id or full path (e.g. "myteam" or "org/myteam").
                   The tool watches every member of this group.
  GITLAB_USERS     Comma-separated usernames to watch instead of / in addition
                   to the group (e.g. "aziz,dilnoza,sardor").
  POLL_INTERVAL    Seconds between refreshes (default 30).
  EVENT_LIMIT      Max rows shown in the feed (default 40).
  DISCORD_WEBHOOK_URL  Optional. If set, every NEW event (after the first poll)
                   is also pushed to this Discord channel webhook.

You must set GITLAB_URL + GITLAB_TOKEN, and at least one of GITLAB_GROUP / GITLAB_USERS.
"""

import os
import sys
import json
import time
from datetime import datetime, timezone
from urllib.parse import urlencode, quote
from urllib.request import Request, urlopen
from urllib.error import HTTPError, URLError

try:
    from rich.console import Console, Group
    from rich.live import Live
    from rich.table import Table
    from rich.panel import Panel
    from rich.text import Text
except ImportError:
    sys.exit("Missing dependency 'rich'. Install with:  pip install rich")


# ---------------------------------------------------------------------------
# Config
# ---------------------------------------------------------------------------

def load_dotenv():
    """Minimal .env loader (KEY=VALUE per line) — no external dependency."""
    path = os.path.join(os.path.dirname(os.path.abspath(__file__)), ".env")
    if not os.path.exists(path):
        return
    with open(path) as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#") or "=" not in line:
                continue
            key, _, val = line.partition("=")
            key, val = key.strip(), val.strip().strip('"').strip("'")
            os.environ.setdefault(key, val)


load_dotenv()

BASE = os.environ.get("GITLAB_URL", "").rstrip("/")
TOKEN = os.environ.get("GITLAB_TOKEN", "")
GROUP = os.environ.get("GITLAB_GROUP", "").strip()
USERS = [u.strip() for u in os.environ.get("GITLAB_USERS", "").split(",") if u.strip()]
POLL_INTERVAL = int(os.environ.get("POLL_INTERVAL", "30"))
EVENT_LIMIT = int(os.environ.get("EVENT_LIMIT", "40"))
DISCORD_WEBHOOK_URL = os.environ.get("DISCORD_WEBHOOK_URL", "").strip()

console = Console()


def fail(msg):
    console.print(f"[bold red]Error:[/] {msg}")
    sys.exit(1)


if not BASE:
    fail("GITLAB_URL is not set. See the README / .env.example.")
if not TOKEN:
    fail("GITLAB_TOKEN is not set. Create a personal access token with `read_api` scope.")
if not GROUP and not USERS:
    fail("Set GITLAB_GROUP and/or GITLAB_USERS so the tool knows whom to watch.")


# ---------------------------------------------------------------------------
# GitLab API
# ---------------------------------------------------------------------------

def api_get(path, params=None):
    url = f"{BASE}/api/v4{path}"
    if params:
        url += "?" + urlencode(params)
    req = Request(url, headers={"PRIVATE-TOKEN": TOKEN, "User-Agent": "gitlab-team-watch"})
    try:
        with urlopen(req, timeout=25) as resp:
            return json.load(resp)
    except HTTPError as e:
        if e.code == 401:
            fail("401 Unauthorized — check GITLAB_TOKEN (needs `read_api` scope).")
        if e.code == 404:
            raise  # let caller decide
        raise RuntimeError(f"HTTP {e.code} on {path}: {e.read().decode(errors='replace')[:200]}")
    except URLError as e:
        raise RuntimeError(f"Cannot reach {BASE} — are you on the office network/VPN? ({e.reason})")


def resolve_members():
    """Return list of {id, name, username} for everyone we should watch."""
    members = {}

    if GROUP:
        gid = quote(GROUP, safe="") if not GROUP.isdigit() else GROUP
        page = 1
        while True:
            try:
                batch = api_get(f"/groups/{gid}/members/all",
                                {"per_page": 100, "page": page})
            except HTTPError:
                fail(f"Group '{GROUP}' not found or not accessible with this token.")
            if not batch:
                break
            for m in batch:
                members[m["id"]] = {"id": m["id"], "name": m["name"],
                                    "username": m["username"]}
            if len(batch) < 100:
                break
            page += 1

    for username in USERS:
        found = api_get("/users", {"username": username})
        if found:
            u = found[0]
            members[u["id"]] = {"id": u["id"], "name": u["name"],
                                "username": u["username"]}
        else:
            console.print(f"[yellow]Warning:[/] user '{username}' not found, skipping.")

    if not members:
        fail("No members resolved. Check GITLAB_GROUP / GITLAB_USERS.")
    return list(members.values())


def fetch_user_events(user, per_page=30):
    """Recent events for a single user, newest first."""
    try:
        events = api_get(f"/users/{user['id']}/events",
                         {"per_page": per_page, "sort": "desc"})
    except HTTPError:
        return []
    for e in events:
        e["_user"] = user  # remember whom we polled (author field is also present)
    return events


# ---------------------------------------------------------------------------
# Rendering
# ---------------------------------------------------------------------------

ACTION_STYLE = {
    "pushed to": "cyan",
    "pushed new": "cyan",
    "deleted": "red",
    "opened": "green",
    "closed": "yellow",
    "merged": "magenta",
    "accepted": "magenta",
    "commented on": "blue",
    "joined": "dim",
    "left": "dim",
    "created": "green",
}


def rel_time(iso):
    try:
        dt = datetime.fromisoformat(iso.replace("Z", "+00:00"))
    except (ValueError, AttributeError):
        return "?"
    delta = datetime.now(timezone.utc) - dt
    s = int(delta.total_seconds())
    if s < 60:
        return f"{s}s"
    if s < 3600:
        return f"{s // 60}m"
    if s < 86400:
        return f"{s // 3600}h"
    return f"{s // 86400}d"


def describe(event):
    """Human-readable description of what happened."""
    action = event.get("action_name", "did")
    target_type = event.get("target_type") or ""
    title = event.get("target_title") or ""
    iid = event.get("target_iid")

    push = event.get("push_data")
    if push:
        count = push.get("commit_count", 0)
        ref = push.get("ref", "")
        verb = "pushed"
        plural = "commit" if count == 1 else "commits"
        ct = push.get("commit_title") or ""
        base = f"{verb} {count} {plural} to {ref}"
        return f"{base} — {ct}" if ct else base

    label = {"MergeRequest": "MR", "Issue": "issue", "Note": "comment",
             "DiffNote": "comment", "Milestone": "milestone",
             "WikiPage::Meta": "wiki"}.get(target_type, target_type.lower())

    ref = f" !{iid}" if target_type == "MergeRequest" and iid else (
          f" #{iid}" if target_type == "Issue" and iid else "")

    parts = [action]
    if label:
        parts.append(label + ref)
    desc = " ".join(parts).strip()
    if title and label not in ("comment",):
        desc += f": {title}"
    return desc


def project_name(event, project_cache):
    pid = event.get("project_id")
    if pid is None:
        return ""
    if pid not in project_cache:
        try:
            proj = api_get(f"/projects/{pid}")
            project_cache[pid] = proj.get("path_with_namespace", str(pid))
        except Exception:
            project_cache[pid] = str(pid)
    return project_cache[pid]


def build_table(events, new_ids, project_cache):
    table = Table(expand=True, show_edge=False, pad_edge=False, box=None)
    table.add_column("", width=2, no_wrap=True)          # new marker
    table.add_column("Who", style="bold", width=16, no_wrap=True)
    table.add_column("Action", ratio=3)
    table.add_column("Project", style="dim", ratio=1, no_wrap=True)
    table.add_column("When", justify="right", width=6, style="dim", no_wrap=True)

    for e in events[:EVENT_LIMIT]:
        author = (e.get("author") or {}).get("name") or e["_user"]["name"]
        action = e.get("action_name", "")
        style = ACTION_STYLE.get(action, "white")
        is_new = e.get("id") in new_ids
        marker = Text("●", style="bold green") if is_new else Text(" ")
        who = Text(author, style="bold white" if is_new else "bold")
        table.add_row(
            marker,
            who,
            Text(describe(e), style=style),
            project_name(e, project_cache),
            rel_time(e.get("created_at", "")),
        )
    return table


# ---------------------------------------------------------------------------
# Discord push notifications (optional)
# ---------------------------------------------------------------------------

EMOJI = {
    "pushed to": "⬆️", "pushed new": "🌱", "deleted": "🗑️",
    "opened": "✨", "closed": "🚪", "merged": "🟣", "accepted": "🟣",
    "commented on": "💬", "joined": "👋", "left": "👋", "created": "✨",
}


def discord_line(event, project_cache):
    """One-line plain-text summary of an event for Discord."""
    author = (event.get("author") or {}).get("name") or event["_user"]["name"]
    action = event.get("action_name", "")
    icon = EMOJI.get(action, "•")
    proj = project_name(event, project_cache)
    when = rel_time(event.get("created_at", ""))
    proj_part = f" · `{proj}`" if proj else ""
    # describe() can repeat the action verb; that's fine and reads naturally.
    return f"{icon} **{author}** {describe(event)}{proj_part} · _{when} ago_"


def post_to_discord(content):
    """POST a message to the Discord channel webhook. Best-effort; never raises."""
    if not DISCORD_WEBHOOK_URL:
        return
    data = json.dumps({"content": content[:1990],
                       "allowed_mentions": {"parse": []}}).encode()
    req = Request(DISCORD_WEBHOOK_URL, data=data,
                  headers={"Content-Type": "application/json",
                           "User-Agent": "gitlab-team-watch"},
                  method="POST")
    try:
        with urlopen(req, timeout=15):
            pass
    except (HTTPError, URLError):
        # Don't let a notification failure kill the live feed.
        pass


def notify_discord(new_events, project_cache):
    """Push new events to Discord, batched to respect the 2000-char message cap."""
    if not DISCORD_WEBHOOK_URL or not new_events:
        return
    # Oldest first so the channel reads chronologically.
    lines = [discord_line(e, project_cache) for e in reversed(new_events)]
    chunk, size = [], 0
    for line in lines:
        if size + len(line) + 1 > 1900 and chunk:
            post_to_discord("\n".join(chunk))
            chunk, size = [], 0
        chunk.append(line)
        size += len(line) + 1
    if chunk:
        post_to_discord("\n".join(chunk))


# ---------------------------------------------------------------------------
# Main loop
# ---------------------------------------------------------------------------

def main():
    console.print(f"[dim]Resolving team members from "
                  f"{'group ' + GROUP if GROUP else ''}"
                  f"{' + ' if GROUP and USERS else ''}"
                  f"{('users ' + ','.join(USERS)) if USERS else ''} …[/]")
    members = resolve_members()
    console.print(f"[green]Watching {len(members)} member(s).[/] "
                  f"Polling every {POLL_INTERVAL}s. Ctrl-C to quit.")
    if DISCORD_WEBHOOK_URL:
        console.print("[cyan]Discord notifications enabled[/] "
                      "(first poll primes the feed; pings start with the next new event).")
    console.print()

    seen = set()            # all event ids we've ever shown
    project_cache = {}      # project_id -> path
    feed = []               # combined event list, newest first

    def poll():
        nonlocal feed
        collected = []
        for m in members:
            collected.extend(fetch_user_events(m))
        # de-dupe by id, keep newest
        by_id = {}
        for e in collected + feed:
            eid = e.get("id")
            if eid is not None:
                by_id[eid] = e
        merged = sorted(by_id.values(),
                        key=lambda e: e.get("created_at", ""), reverse=True)
        new_ids = {e["id"] for e in merged if e["id"] not in seen}
        seen.update(new_ids)
        feed = merged[:max(EVENT_LIMIT * 3, 120)]
        return new_ids

    def render(new_ids, status):
        header = Text(status, style="dim")
        table = build_table(feed, new_ids, project_cache)
        return Panel(Group(header, Text(""), table),
                     title="[bold]GitLab Team Activity[/]",
                     subtitle=f"[dim]{len(members)} members · {BASE}[/]",
                     border_style="cyan")

    first_poll = True
    try:
        with Live(console=console, screen=True, refresh_per_second=4) as live:
            while True:
                stamp = datetime.now().strftime("%H:%M:%S")
                live.update(render(set(), f"Refreshing… (last attempt {stamp})"))
                try:
                    new_ids = poll()
                    # Skip the first poll's "new" set — it's the whole backlog, not
                    # genuinely new activity — so we don't spam Discord at startup.
                    if not first_poll:
                        notify_discord([e for e in feed if e["id"] in new_ids],
                                       project_cache)
                    first_poll = False
                    status = (f"Updated {datetime.now().strftime('%H:%M:%S')} · "
                              f"{len(new_ids)} new since last poll · "
                              f"next in {POLL_INTERVAL}s")
                except RuntimeError as err:
                    new_ids = set()
                    status = f"[red]{err}[/] · retrying in {POLL_INTERVAL}s"
                live.update(render(new_ids, status))
                time.sleep(POLL_INTERVAL)
    except KeyboardInterrupt:
        console.print("\n[dim]Stopped.[/]")


if __name__ == "__main__":
    main()

File: gitlab-watch.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');

const C = {
  reset: '\x1b[0m', bold: '\x1b[1m', dim: '\x1b[2m',
  red: '\x1b[31m', green: '\x1b[32m', yellow: '\x1b[33m',
  blue: '\x1b[34m', magenta: '\x1b[35m', cyan: '\x1b[36m', white: '\x1b[37m',
};
const color = (s, ...codes) => codes.map((c) => C[c]).join('') + s + C.reset;

function loadDotenv() {
  const p = path.join(__dirname, '.env');
  if (!fs.existsSync(p)) return;
  for (const raw of fs.readFileSync(p, 'utf8').split('\n')) {
    const line = raw.trim();
    if (!line || line.startsWith('#') || !line.includes('=')) continue;
    const idx = line.indexOf('=');
    const key = line.slice(0, idx).trim();
    let val = line.slice(idx + 1).trim();
    if ((val.startsWith('"') && val.endsWith('"')) ||
        (val.startsWith("'") && val.endsWith("'"))) {
      val = val.slice(1, -1);
    }
    if (!(key in process.env)) process.env[key] = val;
  }
}

loadDotenv();

const BASE = (process.env.GITLAB_URL || '').replace(/\/+$/, '');
const TOKEN = process.env.GITLAB_TOKEN || '';
const GROUP = (process.env.GITLAB_GROUP || '').trim();
const USERS = (process.env.GITLAB_USERS || '').split(',').map((u) => u.trim()).filter(Boolean);
const POLL_INTERVAL = parseInt(process.env.POLL_INTERVAL || '30', 10);
const EVENT_LIMIT = parseInt(process.env.EVENT_LIMIT || '40', 10);
const DISCORD_WEBHOOK_URL = (process.env.DISCORD_WEBHOOK_URL || '').trim();

function fail(msg) {
  process.stderr.write(color(`Error: `, 'bold', 'red') + msg + '\n');
  process.exit(1);
}

if (!BASE) fail('GITLAB_URL is not set. See the README / .env.example.');
if (!TOKEN) fail('GITLAB_TOKEN is not set. Create a personal access token with `read_api` scope.');
if (!GROUP && USERS.length === 0) {
  fail('Set GITLAB_GROUP and/or GITLAB_USERS so the tool knows whom to watch.');
}

class HttpError extends Error {
  constructor(status, body) {
    super(`HTTP ${status}`);
    this.status = status;
    this.body = body;
  }
}

async function apiGet(p, params) {
  let url = `${BASE}/api/v4${p}`;
  if (params) url += '?' + new URLSearchParams(params).toString();
  let resp;
  try {
    resp = await fetch(url, {
      headers: { 'PRIVATE-TOKEN': TOKEN, 'User-Agent': 'gitlab-team-watch' },
      signal: AbortSignal.timeout(25000),
    });
  } catch (e) {
    throw new Error(`Cannot reach ${BASE} — are you on the office network/VPN? (${e.message})`);
  }
  if (!resp.ok) {
    if (resp.status === 401) fail('401 Unauthorized — check GITLAB_TOKEN (needs `read_api` scope).');
    const body = (await resp.text()).slice(0, 200);
    throw new HttpError(resp.status, body);
  }
  return resp.json();
}

async function resolveMembers() {
  const members = new Map();

  if (GROUP) {
    const gid = /^\d+$/.test(GROUP) ? GROUP : encodeURIComponent(GROUP);
    let page = 1;
    for (;;) {
      let batch;
      try {
        batch = await apiGet(`/groups/${gid}/members/all`, { per_page: '100', page: String(page) });
      } catch (e) {
        if (e instanceof HttpError) fail(`Group '${GROUP}' not found or not accessible with this token.`);
        throw e;
      }
      if (!batch.length) break;
      for (const m of batch) members.set(m.id, { id: m.id, name: m.name, username: m.username });
      if (batch.length < 100) break;
      page += 1;
    }
  }

  for (const username of USERS) {
    const found = await apiGet('/users', { username });
    if (found.length) {
      const u = found[0];
      members.set(u.id, { id: u.id, name: u.name, username: u.username });
    } else {
      process.stderr.write(color('Warning: ', 'yellow') + `user '${username}' not found, skipping.\n`);
    }
  }

  if (members.size === 0) fail('No members resolved. Check GITLAB_GROUP / GITLAB_USERS.');
  return [...members.values()];
}

async function fetchUserEvents(user, perPage = 30) {
  let events;
  try {
    events = await apiGet(`/users/${user.id}/events`, { per_page: String(perPage), sort: 'desc' });
  } catch (e) {
    if (e instanceof HttpError) return [];
    throw e;
  }
  for (const e of events) e._user = user;
  return events;
}

const ACTION_COLOR = {
  'pushed to': 'cyan', 'pushed new': 'cyan', deleted: 'red',
  opened: 'green', closed: 'yellow', merged: 'magenta', accepted: 'magenta',
  'commented on': 'blue', joined: 'dim', left: 'dim', created: 'green',
};

const TARGET_LABEL = {
  MergeRequest: 'MR', Issue: 'issue', Note: 'comment', DiffNote: 'comment',
  Milestone: 'milestone', 'WikiPage::Meta': 'wiki',
};

function relTime(iso) {
  const t = Date.parse(iso);
  if (Number.isNaN(t)) return '?';
  const s = Math.floor((Date.now() - t) / 1000);
  if (s < 60) return `${s}s`;
  if (s < 3600) return `${Math.floor(s / 60)}m`;
  if (s < 86400) return `${Math.floor(s / 3600)}h`;
  return `${Math.floor(s / 86400)}d`;
}

function describe(event) {
  const action = event.action_name || 'did';
  const targetType = event.target_type || '';
  const title = event.target_title || '';
  const iid = event.target_iid;

  const push = event.push_data;
  if (push) {
    const count = push.commit_count || 0;
    const ref = push.ref || '';
    const plural = count === 1 ? 'commit' : 'commits';
    const ct = push.commit_title || '';
    const base = `pushed ${count} ${plural} to ${ref}`;
    return ct ? `${base} — ${ct}` : base;
  }

  const label = TARGET_LABEL[targetType] ?? targetType.toLowerCase();
  let ref = '';
  if (targetType === 'MergeRequest' && iid) ref = ` !${iid}`;
  else if (targetType === 'Issue' && iid) ref = ` #${iid}`;

  const parts = [action];
  if (label) parts.push(label + ref);
  let desc = parts.join(' ').trim();
  if (title && label !== 'comment') desc += `: ${title}`;
  return desc;
}

const projectCache = new Map();
async function projectName(event) {
  const pid = event.project_id;
  if (pid == null) return '';
  if (!projectCache.has(pid)) {
    try {
      const proj = await apiGet(`/projects/${pid}`);
      projectCache.set(pid, proj.path_with_namespace || String(pid));
    } catch {
      projectCache.set(pid, String(pid));
    }
  }
  return projectCache.get(pid);
}

function fit(s, width, align = 'left') {
  s = String(s ?? '');
  if (s.length > width) s = width > 1 ? s.slice(0, width - 1) + '…' : s.slice(0, width);
  return align === 'right' ? s.padStart(width) : s.padEnd(width);
}

async function buildRows(events, newIds) {
  const cols = process.stdout.columns || 100;
  const wMark = 2, wWho = 16, wWhen = 6, wProj = Math.min(24, Math.max(10, Math.floor(cols * 0.2)));
  const wAction = Math.max(20, cols - wMark - wWho - wProj - wWhen - 4);

  const lines = [];
  for (const e of events.slice(0, EVENT_LIMIT)) {
    const author = (e.author && e.author.name) || e._user.name;
    const action = e.action_name || '';
    const col = ACTION_COLOR[action] || 'white';
    const isNew = newIds.has(e.id);

    const mark = isNew ? color('*', 'bold', 'green') : ' '.repeat(wMark);
    const who = color(fit(author, wWho), 'bold');
    const act = color(fit(describe(e), wAction), col);
    const proj = color(fit(await projectName(e), wProj), 'dim');
    const when = color(fit(relTime(e.created_at || ''), wWhen, 'right'), 'dim');
    lines.push(`${fit(mark, wMark)} ${who} ${act} ${proj} ${when}`);
  }
  return lines;
}

const CLEAR = '\x1b[2J\x1b[H';

async function render(events, newIds, status, memberCount) {
  const rows = await buildRows(events, newIds);
  const title = color(' GitLab Team Activity ', 'bold', 'cyan');
  const sub = color(`${memberCount} members · ${BASE}`, 'dim');
  let out = CLEAR;
  out += title + '\n';
  out += color(status, 'dim') + '\n\n';
  out += rows.join('\n') + '\n\n';
  out += sub + '\n';
  process.stdout.write(out);
}

function clockTime(iso) {
  const t = Date.parse(iso);
  if (Number.isNaN(t)) return '--:--:--';
  return new Date(t).toLocaleTimeString('en-GB', { hour12: false });
}

async function discordLine(event) {
  const author = (event.author && event.author.name) || event._user.name;
  const proj = await projectName(event);
  const time = clockTime(event.created_at || '');
  const projPart = proj ? ` · \`${proj}\`` : '';
  return `\`${time}\`  **${author}**  ${describe(event)}${projPart}`;
}

async function postToDiscord(content) {
  if (!DISCORD_WEBHOOK_URL) return;
  try {
    await fetch(DISCORD_WEBHOOK_URL, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json', 'User-Agent': 'gitlab-team-watch' },
      body: JSON.stringify({ content: content.slice(0, 1990), allowed_mentions: { parse: [] } }),
      signal: AbortSignal.timeout(15000),
    });
  } catch {
  }
}

async function notifyDiscord(newEvents) {
  if (!DISCORD_WEBHOOK_URL || newEvents.length === 0) return;

  const ordered = [...newEvents].reverse();
  const lines = [];
  for (const e of ordered) lines.push(await discordLine(e));

  const n = ordered.length;
  const title = `**${n} new event${n === 1 ? '' : 's'}**`;

  let first = true;
  let chunk = [], size = title.length + 1;
  for (const line of lines) {
    if (size + line.length + 1 > 1900 && chunk.length) {
      await postToDiscord((first ? title + '\n' : '') + chunk.join('\n'));
      first = false; chunk = []; size = 0;
    }
    chunk.push(line);
    size += line.length + 1;
  }
  if (chunk.length) await postToDiscord((first ? title + '\n' : '') + chunk.join('\n'));
}

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function main() {
  const watching = [
    GROUP ? `group ${GROUP}` : '',
    GROUP && USERS.length ? ' + ' : '',
    USERS.length ? `users ${USERS.join(',')}` : '',
  ].join('');
  process.stdout.write(color(`Resolving team members from ${watching} …\n`, 'dim'));

  const members = await resolveMembers();
  process.stdout.write(color(`Watching ${members.length} member(s).`, 'green') +
    ` Polling every ${POLL_INTERVAL}s. Ctrl-C to quit.\n`);
  if (DISCORD_WEBHOOK_URL) {
    process.stdout.write(color('Discord notifications enabled', 'cyan') +
      ' (first poll primes the feed; pings start with the next new event).\n');
  }

  const seen = new Set();
  let feed = [];
  let firstPoll = true;

  async function poll() {
    let collected = [];
    for (const m of members) collected = collected.concat(await fetchUserEvents(m));
    const byId = new Map();
    for (const e of [...collected, ...feed]) {
      if (e.id != null) byId.set(e.id, e);
    }
    const merged = [...byId.values()].sort((a, b) =>
      String(b.created_at || '').localeCompare(String(a.created_at || '')));
    const newIds = new Set(merged.filter((e) => !seen.has(e.id)).map((e) => e.id));
    for (const id of newIds) seen.add(id);
    feed = merged.slice(0, Math.max(EVENT_LIMIT * 3, 120));
    return newIds;
  }

  process.stdout.write('\x1b[?1049h\x1b[?25l');
  const restore = () => process.stdout.write('\x1b[?25h\x1b[?1049l');
  process.on('SIGINT', () => { restore(); process.stdout.write(color('Stopped.\n', 'dim')); process.exit(0); });

  for (;;) {
    const stamp = new Date().toLocaleTimeString();
    await render(feed, new Set(), `Refreshing… (last attempt ${stamp})`, members.length);
    let newIds, status;
    try {
      newIds = await poll();
      if (!firstPoll) await notifyDiscord(feed.filter((e) => newIds.has(e.id)));
      firstPoll = false;
      status = `Updated ${new Date().toLocaleTimeString()} · ${newIds.size} new since last poll · next in ${POLL_INTERVAL}s`;
    } catch (err) {
      newIds = new Set();
      status = `${err.message} · retrying in ${POLL_INTERVAL}s`;
    }
    await render(feed, newIds, status, members.length);
    await sleep(POLL_INTERVAL * 1000);
  }
}

main().catch((e) => { process.stdout.write('\x1b[?25h\x1b[?1049l'); fail(e.message); });
